File: test/go/widget/apply_ott_test.go
package widget_test

import (
	"context"
	"encoding/json"
	"fmt"
	"testing"
	"uat-script/helper"
	widget_helper "uat-script/widget"

	"github.com/dana-id/dana-go/v2/widget/v1"
)

const (
	widgetApplyOttCase = "ApplyOtt"
)

// ApplyOtt
func TestApplyOttSuccess(t *testing.T) {
	helper.RetryTest(t, 3, 1, func() error {
		caseName := "ApplyOttSuccess"

		// Get auth code via OAuth flow
		redirectUrlAuthCode, _ := widget_helper.GetRedirectOauthUrl(
			helper.TestConfig.PhoneNumber,
			helper.TestConfig.PIN,
		)
		authCode, _ := widget_helper.GetAuthCode(
			helper.TestConfig.PhoneNumber,
			helper.TestConfig.PIN,
			redirectUrlAuthCode,
		)

		// Exchange auth code for a real access token via ApplyToken API
		ctx := context.Background()
		applyTokenReq := widget.NewApplyTokenAuthorizationCodeRequest("AUTHORIZATION_CODE", authCode)
		applyTokenReqValue := widget.ApplyTokenAuthorizationCodeRequestAsApplyTokenRequest(applyTokenReq)
		applyTokenResp, _, err := helper.ApiClient.WidgetAPI.ApplyToken(ctx).ApplyTokenRequest(applyTokenReqValue).Execute()
		if err != nil {
			return fmt.Errorf("failed to obtain access token: %v", err)
		}
		accessToken := applyTokenResp.GetAccessToken()
		fmt.Printf("Obtained access token: %s\n", accessToken)

		// Get the request data from JSON and only patch accessToken (matching PHP)
		jsonDict, err := helper.GetRequest(helper.TestConfig.JsonWidgetPath, widgetApplyOttCase, caseName)
		if err != nil {
			return fmt.Errorf("failed to get request data: %v", err)
		}

		if additionalInfo, ok := jsonDict["additionalInfo"].(map[string]interface{}); ok {
			additionalInfo["accessToken"] = accessToken
		} else {
			jsonDict["additionalInfo"] = map[string]interface{}{"accessToken": accessToken}
		}

		jsonBytes, err := json.Marshal(jsonDict)
		if err != nil {
			return fmt.Errorf("failed to marshal JSON: %v", err)
		}

		applyOttReq := widget.ApplyOTTRequest{}
		err = json.Unmarshal(jsonBytes, &applyOttReq)
		if err != nil {
			return fmt.Errorf("failed to unmarshal JSON: %v", err)
		}

		apiResponse, httpResponse, err := helper.ApiClient.WidgetAPI.ApplyOTT(ctx).ApplyOTTRequest(applyOttReq).Execute()
		if err != nil {
			return fmt.Errorf("API request failed: %v", err)
		}
		defer httpResponse.Body.Close()

		responseJSON, err := apiResponse.MarshalJSON()
		if err != nil {
			return fmt.Errorf("failed to convert response to JSON: %v", err)
		}

		err = helper.AssertResponse(
			helper.TestConfig.JsonWidgetPath,
			widgetApplyOttCase,
			caseName,
			string(responseJSON),
			nil,
		)
		if err != nil {
			return err
		}
		return nil
	})
}

func TestApplyOttFailTokenNotFound(t *testing.T) {
	caseName := "ApplyOttCustomerTokenNotFound"
	jsonDict, err := helper.GetRequest(helper.TestConfig.JsonWidgetPath, widgetApplyOttCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}

	// Only patch accessToken (matching PHP: $jsonDict['additionalInfo']['accessToken'] = 'invalid_access_token_for_testing')
	if additionalInfo, ok := jsonDict["additionalInfo"].(map[string]interface{}); ok {
		additionalInfo["accessToken"] = "invalid_access_token_for_testing"
	} else {
		jsonDict["additionalInfo"] = map[string]interface{}{"accessToken": "invalid_access_token_for_testing"}
	}

	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	applyOttReq := widget.ApplyOTTRequest{}
	err = json.Unmarshal(jsonBytes, &applyOttReq)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	ctx := context.Background()
	_, httpResponse, err := helper.ApiClient.WidgetAPI.ApplyOTT(ctx).ApplyOTTRequest(applyOttReq).Execute()
	if err != nil {
		err = helper.AssertFailResponse(helper.TestConfig.JsonWidgetPath, widgetApplyOttCase, caseName, httpResponse, nil)
		if err != nil {
			t.Fatal(err)
		}
	} else {
		defer httpResponse.Body.Close()
		t.Fatalf("Expected ApiException was not thrown")
	}
}

func TestApplyOttFailInvalidUserStatus(t *testing.T) {
	t.Skip("Scenario skipped because the result the same with testApplyOttFailTokenNotFound")
	caseName := "ApplyOttFailInvalidUserStatus"
	// Get the request data from JSON
	jsonDict, err := helper.GetRequest(helper.TestConfig.JsonWidgetPath, widgetApplyOttCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}
	jsonDict["authCode"] = "GtRLpA0TyqK3becMq4dCMnVf1N9KLHNixVfC1800"

	// Marshal to JSON and unmarshal to widget SDK struct for type safety
	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	applyOTTRequest := &widget.ApplyOTTRequest{}
	err = json.Unmarshal(jsonBytes, applyOTTRequest)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	// Execute the SDK API call with success expectation
	ctx := context.Background()

	// Make the API call using the Widget SDK
	_, httpResponse, err := helper.ApiClient.WidgetAPI.ApplyOTT(ctx).ApplyOTTRequest(*applyOTTRequest).Execute()
	if err != nil {
		// Assert the error response matches expected error pattern
		err = helper.AssertFailResponse(helper.TestConfig.JsonWidgetPath, widgetApplyOttCase, caseName, httpResponse, nil)
		if err != nil {
			t.Fatal(err)
		}
	} else {
		// If no error occurred, this is unexpected for error test cases
		defer httpResponse.Body.Close()
		t.Fatalf("Expected error for case %s but API call succeeded", caseName)
	}
}

File: test/node/widget/apply_ott_test.ts
import Dana, { ResponseError } from 'dana-node';
import { v4 as uuidv4 } from 'uuid';
import * as path from 'path';
import * as dotenv from 'dotenv';
import { fail } from 'assert';
import { getRequest } from '../helper/util';
import { assertResponse, assertFailResponse } from '../helper/assertion';
import { ApplyOTTRequest, ApplyOTTRequestAdditionalInfo, ApplyTokenRequest } from 'dana-node/widget/v1';
import { executeManualApiRequest } from '../helper/apiHelpers';
import { generateAuthCode } from '../helper/auth-utils';

dotenv.config();

const titleCase = 'ApplyOtt';
const jsonPathFile = path.resolve(__dirname, '../../../resource/request/components/Widget.json');

const dana = new Dana({
    partnerId: process.env.X_PARTNER_ID || '',
    privateKey: process.env.PRIVATE_KEY || '',
    origin: process.env.ORIGIN || '',
    env: process.env.ENV || 'sandbox',
});

/**
 * Generate an access token for testing using an auth code from the helper utility
 * @param phoneNumber Optional phone number for OAuth flow
 * @param pinCode Optional PIN code for OAuth flow
 * @returns Promise resolving to access token string
 */
async function generateApplyToken(phoneNumber?: string, pinCode?: string): Promise<string> {
    const caseName = 'ApplyTokenSuccess';
    const requestData: ApplyTokenRequest = getRequest<ApplyTokenRequest>(jsonPathFile, "ApplyToken", caseName);

    requestData.authCode = await generateAuthCode(phoneNumber, pinCode);

    const response = await dana.widgetApi.applyToken(requestData);
    return response.accessToken;
}

describe('ApplyOtt Tests', () => {
    test('should successfully apply OTT', async () => {
        const caseName = 'ApplyOttSuccess';
        const requestData: ApplyOTTRequest = getRequest<ApplyOTTRequest>(jsonPathFile, titleCase, caseName);

        requestData.additionalInfo.accessToken = await generateApplyToken();
        try {
            const response = await dana.widgetApi.applyOTT(requestData);
            await assertResponse(jsonPathFile, titleCase, caseName, response);
        } catch (e: any) {
            fail('ApplyOTT test failed: ' + (e.message || e));
        }
    });

    test('should fail to apply OTT with invalid format', async () => {
        const caseName = 'ApplyOttFailInvalidFormat';
        const requestData: ApplyOTTRequest = getRequest<ApplyOTTRequest>(jsonPathFile, titleCase, caseName);
        requestData.additionalInfo.accessToken = await generateApplyToken();

        try {
            await dana.widgetApi.applyOTT(requestData);
            fail('Expected an error but the API call succeeded');
        } catch (e: any) {
            if (e instanceof ResponseError && Number(e.status) === 400) {
                // Expected error for invalid format
                await assertFailResponse(jsonPathFile, titleCase, caseName, e.rawResponse);
            } else if (e instanceof ResponseError && Number(e.status) === 401) {
                // Expected error for invalid signature
                fail("Expected unauthorized failed but got status code " + e.status);
            } else {
                // Unexpected error
                console.error('Unexpected error:', e);
            }
        }

    });

    test('should fail to apply OTT with missing or invalid mandatory field', async () => {
        const caseName = 'ApplyOttFailMissingOrInvalidMandatoryField';
        try {
            const requestData: ApplyOTTRequest = getRequest<ApplyOTTRequest>(jsonPathFile, titleCase, caseName);

            const baseUrl: string = 'https://api.sandbox.dana.id/';
            const apiPath: string = '/rest/v1.1/qr/apply-ott';
            const customHeaders: Record<string, string> = {
                'X-TIMESTAMP': ""
            };

            await executeManualApiRequest(
                caseName,
                "POST",
                baseUrl + apiPath,
                apiPath,
                requestData,
                customHeaders,
            );
            fail('Expected an error but the API call succeeded');
        } catch (e: any) {
            if (Number(e.status) === 400) {
                // Expected error for invalid format
                await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse));
            }
            else if (e instanceof ResponseError) {
                // Expected error for invalid signature
                fail("Expected unauthorized failed but got status code " + e.status);
            } else {
                throw e;
            }
        }
    });

    test('should fail to apply OTT with invalid signature', async () => {
        const caseName = 'ApplyOttFailInvalidSignature';
        const requestData: ApplyOTTRequest = getRequest<ApplyOTTRequest>(jsonPathFile, titleCase, caseName);
        requestData.additionalInfo.accessToken = await generateApplyToken();
        // Custom headers: use invalid signature to trigger authorization error


        try {
            const baseUrl: string = 'https://api.sandbox.dana.id/';
            const apiPath: string = '/rest/v1.1/qr/apply-ott';

            const customHeaders: Record<string, string> = {
                'Authorization-Customer': 'Bearer ' + requestData.additionalInfo.accessToken,
                'X-DEVICE-ID': '1234567890',
                'X-SIGNATURE': '85be817c55b2c135157c7e89f52499bf0c25ad6eeebe04a986e8c862561b19a5'
            };
            // Make direct API call with invalid signature
            await executeManualApiRequest(
                caseName,
                "POST",
                baseUrl + apiPath,
                apiPath,
                requestData,
                customHeaders
            );

            fail("Expected an error but the API call succeeded");
        } catch (e: any) {
            // Expecting a 401 Unauthorized error
            if (Number(e.status) === 401) {
                await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse));
            } else if (e instanceof ResponseError && Number(e.status) !== 401) {
                fail("Expected unauthorized failed but got status code " + e.status + JSON.stringify(e.rawResponse));
            } else {
                throw e;
            }
        }
    });

    test('should fail to apply OTT with token expired', async () => {
        const caseName = 'ApplyOttFailTokenExpired';
        const requestData: ApplyOTTRequest = getRequest<ApplyOTTRequest>(jsonPathFile, titleCase, caseName);

        requestData.additionalInfo.accessToken = await generateApplyToken("08551003634", "574008");

        try {
            const response = await dana.widgetApi.applyOTT(requestData);
            await assertFailResponse(jsonPathFile, titleCase, caseName, response);
            fail('Expected an error but the API call succeeded');
        } catch (e: any) { }
    });

    test('should fail to apply OTT with token not found', async () => {
        const caseName = 'ApplyOttFailTokenNotFound';
        const requestData: ApplyOTTRequest = getRequest<ApplyOTTRequest>(jsonPathFile, titleCase, caseName);
        requestData.additionalInfo.accessToken = uuidv4(); // Use a random token to simulate not found
        try {
            const response = await dana.widgetApi.applyOTT(requestData);
            await assertFailResponse(jsonPathFile, titleCase, caseName, response);
            fail('Expected an error but the API call succeeded');
        } catch (e: any) { }
    });

    test.skip('should fail to apply OTT with invalid user status', async () => {
        const caseName = 'ApplyOttFailInvalidUserStatus';
        const requestData: ApplyOTTRequest = getRequest<ApplyOTTRequest>(jsonPathFile, titleCase, caseName);
        requestData.additionalInfo.accessToken = await generateApplyToken("0855100800", "146838");
        try {
            const response = await dana.widgetApi.applyOTT(requestData);
            await assertFailResponse(jsonPathFile, titleCase, caseName, response);
            fail('Expected an error but the API call succeeded');
        } catch (e: any) { }
    });

    test.skip('should fail to apply OTT with non-retryable error', async () => {
        const caseName = 'ApplyOttFailNonRetryableError';
        const requestData: ApplyOTTRequest = getRequest<ApplyOTTRequest>(jsonPathFile, titleCase, caseName);
        requestData.additionalInfo.accessToken = await generateApplyToken("0815919191", "945922");
        try {
            const response = await dana.widgetApi.applyOTT(requestData);
            await assertFailResponse(jsonPathFile, titleCase, caseName, response);
            fail('Expected an error but the API call succeeded');
        } catch (e: any) { }
    });

    test.skip('should fail to apply OTT with internal server error', async () => {
        const caseName = 'ApplyOttFailInternalServerError';
        const requestData: ApplyOTTRequest = getRequest<ApplyOTTRequest>(jsonPathFile, titleCase, caseName);
        requestData.additionalInfo.accessToken = await generateApplyToken("081298055132", "677832");
        try {
            const response = await dana.widgetApi.applyOTT(requestData);
            await assertFailResponse(jsonPathFile, titleCase, caseName, response);
            fail('Expected an error but the API call succeeded');
        } catch (e: any) { }
    });

    test.skip('should fail to apply OTT with unexpected response', async () => {
        const caseName = 'ApplyOttFailUnexpectedResponse';
        const requestData: ApplyOTTRequest = getRequest<ApplyOTTRequest>(jsonPathFile, titleCase, caseName);
        requestData.additionalInfo.accessToken = await generateApplyToken("08121532586", "944279");
        try {
            const response = await dana.widgetApi.applyOTT(requestData);
            await assertFailResponse(jsonPathFile, titleCase, caseName, response);
            fail('Expected an error but the API call succeeded');
        } catch (e: any) { }
    });
});
